crates/bin/prove_block/src/main.rs

```diff
@@ -189,20 +193,62 @@ async fn main() -> Result<(), Box<dyn Error>> {
             }
         };
 
-    let state_update =
-        match provider.get_state_update(BlockId::Number(block_number)).await.expect("Failed to get state update") {
-            MaybePendingStateUpdate::Update(update) => update,
-            MaybePendingStateUpdate::PendingUpdate(_) => {
-                panic!("Block is still pending!")
+    let state_update = match provider.get_state_update(block_id).await.expect("Failed to get state update") {
+        MaybePendingStateUpdate::Update(update) => update,
+        MaybePendingStateUpdate::PendingUpdate(_) => {
+            panic!("Block is still pending!")
+        }
+    };
+
+    // utility to extract all contract address in a nested call structure
+    fn process_function_invocations(inv: FunctionInvocation, contracts: &mut HashSet<Felt252>, nesting_level: u64) {
+        log::info!("process_function_inv: contract: {:x} at level {}", inv.contract_address, nesting_level);
+        contracts.insert(inv.contract_address);
+        for call in inv.calls {
+            process_function_invocations(call, contracts, nesting_level + 1);
+        }
+    }
+    // extract other contracts used in our block from the block trace
+    let mut contracts_subcalled = HashSet::new();
+    let traces = provider.trace_block_transactions(block_id).await.expect("Failed to get block tx traces");
+    for trace in traces {
+        match trace.trace_root {
+            TransactionTrace::Invoke(invoke_trace) => {
+                if let Some(inv) = invoke_trace.validate_invocation {
+                    process_function_invocations(inv, &mut contracts_subcalled, 0);
+                }
+                match invoke_trace.execute_invocation {
+                    ExecuteInvocation::Success(inv) => {
+                        process_function_invocations(inv, &mut contracts_subcalled, 0);
+                    }
+                    ExecuteInvocation::Reverted(_) => unimplemented!("handle reverted invoke trace"),
+                }
+                if let Some(inv) = invoke_trace.fee_transfer_invocation {
+                    process_function_invocations(inv, &mut contracts_subcalled, 0);
+                }
             }
-        };
+            _ => unimplemented!("process other txn traces"),
+        }
+    }
```

Comment:
This will fails as soon as we try with other block. But ok for now :) 
Already created a ticket for Fernando :P